src/shader_recompiler/ir/passes/flatten_extended_userdata_pass.cpp

```diff
@@ -0,0 +1,267 @@
+
+// SPDX-FileCopyrightText: Copyright 2024 shadPS4 Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include <unordered_map>
+#include <boost/container/flat_map.hpp>
+#include <xbyak/xbyak.h>
+#include <xbyak/xbyak_util.h>
+#include "common/config.h"
+#include "common/io_file.h"
+#include "common/path_util.h"
+#include "common/singleton.h"
+#include "shader_recompiler/info.h"
+#include "shader_recompiler/ir/breadth_first_search.h"
+#include "shader_recompiler/ir/opcodes.h"
+#include "shader_recompiler/ir/passes/srt.h"
+#include "shader_recompiler/ir/program.h"
+#include "shader_recompiler/ir/reg.h"
+#include "shader_recompiler/ir/srt_gvn_table.h"
+#include "shader_recompiler/ir/value.h"
+#include "src/common/arch.h"
+#include "src/common/decoder.h"
+
+using namespace Xbyak::util;
+
+// TODO make sure no problems with identity and Insts being used in maps
+
+static void DumpSrtProgram(const Shader::Info& info, const u8* code, size_t codesize) {
+#ifdef ARCH_X86_64
+    using namespace Common::FS;
+
+    const auto dump_dir = GetUserPath(PathType::ShaderDir) / "dumps";
+    if (!std::filesystem::exists(dump_dir)) {
+        std::filesystem::create_directories(dump_dir);
+    }
+    const auto filename =
+        fmt::format("{}_{:#018x}_{}.srtprogram.txt", info.stage, info.pgm_hash, info.perm_idx);
+    const auto file = IOFile{dump_dir / filename, FileAccessMode::Write, FileType::TextFile};
+
+    u64 address = reinterpret_cast<u64>(code);
+    u64 code_end = address + codesize;
+    ZydisDecodedInstruction instruction;
+    ZydisDecodedOperand operands[ZYDIS_MAX_OPERAND_COUNT];
+    ZyanStatus status = ZYAN_STATUS_SUCCESS;
+    while (address < code_end && ZYAN_SUCCESS(Common::Decoder::Instance()->decodeInstruction(
+                                     instruction, operands, reinterpret_cast<void*>(address)))) {
+        std::string s =
+            Common::Decoder::Instance()->disassembleInst(instruction, operands, address);
+        s += "\n";
+        file.WriteString(s);
+        address += instruction.length;
+    }
+#endif
+}
+
+namespace {
+class SrtCodegen : public Xbyak::CodeGenerator {
```

Comment:
This is superfluous abstraction I think, can use Xbyak::CodeGenerator directly. Also it can be constructed each time its used, doesn't need to be a global singleton if memory is managed externally. Or you could use a static Xbyak::Codegen in this file if the memory can be managed by xbyak during the duration of the program (I think easier overall)

Reply:
you can make the call on this. Personally id either use SmallCodeArray or a single Xbyak::CodeGenerator per Shader::Info.
Although these are inneficient since they need a page at minimum per program, and programs need separate allocations for garbage collection purposes.
my last choice would be using a single static Codegenerator that never gets reset cuz of the permutation problem